pkg/agent/manager/sync.go

```diff
@@ -0,0 +1,302 @@
+package manager
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+	"github.com/spiffe/spire/proto/common"
+	proto "github.com/spiffe/spire/proto/common"
+)
+
+// synchronize hits the node api, checks for entries we haven't fetched yet, and fetches them.
+func (m *manager) synchronize(spiffeID string) (err error) {
+	m.c.Log.Debugf("synchronize started for %s", spiffeID)
+	defer m.c.Log.Debug("synchronize finished")
+
+	var regEntries map[string]*proto.RegistrationEntry
+	var cEntryRequests entryRequests
+
+	regEntries, _, err = m.fetchUpdates(spiffeID, nil)
+	if err != nil {
+		return err
+	}
+
+	cEntryRequests, err = m.checkExpiredCacheEntries()
+	if err != nil {
+		return err
+	}
+
+	// While there are registration entries or cache entries to process...
+	for len(regEntries) > 0 || len(cEntryRequests) > 0 {
+		cEntryRequests, err = m.checkForNewCacheEntries(regEntries, cEntryRequests)
+		if err != nil {
+			return err
+		}
+
+		regEntries, err = m.processEntryRequests(cEntryRequests)
+		if err != nil {
+			return err
+		}
+		cEntryRequests = entryRequests{}
+	}
+
+	return nil
+}
+
+// entryRequest holds a CSR and a pre-built cache entry for the RegistrationEntry
+// contained in the entry field.
+type entryRequest struct {
+	CSR   []byte
+	entry *cache.Entry
+}
+
+// entryRequests is a map keyed by SPIFFEID where each value is a list
+// of entryRequest that should be processed using the SVID and PrivateKey
+// created for it.
+type entryRequests map[string][]*entryRequest
+
+func (m *manager) fetchUpdates(spiffeID string, entryRequests []*entryRequest) (map[string]*common.RegistrationEntry, map[string]*node.Svid, error) {
+	m.c.Log.Debugf("fetchUpdates started using SVID and key for spiffeID: %s", spiffeID)
+	defer m.c.Log.Debug("fetchUpdates finished")
+
+	client := m.syncClients.get(spiffeID)
+	if client == nil {
+		return nil, nil, fmt.Errorf("No client found for %s", spiffeID)
```

Comment:
nit: error strings should be lowercase

reference here for idiomatic go error:
https://blog.golang.org/error-handling-and-go
https://golang.org/doc/effective_go.html#errors

Reply:
:+1:  will update them.